### PR TITLE
feat(flightsql): Implement FlightSQL `GetSqlInfo` endpoint

### PR DESCRIPTION
**Pull request by alamb**
closes https://github.com/influxdata/idpe/issues/17197

related to https://github.com/influxdata/influxdb_iox/issues/7210

# Background

The [`GetSqlInfo`](https://github.com/apache/arrow/blob/f1eece9f276184063c9c35011e8243eb3b071233/format/FlightSql.proto#L26-L68
) command in FlightSQL is used to provide metadata to client programs, via APIs such as [JDBC DatabaseMetaData](https://docs.oracle.com/javase/8/docs/api/java/sql/DatabaseMetaData.html). This PR ports the `queryrouterd` implementation [source link](https://github.com/influxdata/idpe/blob/85aa7a52b40f173cc4d79ac02b3a4a13e82333c4/queryrouter/internal/server/flightsql_handler.go#L208-L275)

# Size Apologies
Note there are only about 250 lines of code in this PR -- the rest are tests. If reviewers would prefer I can break it up into smaller PRs for review

# Features:
IOx now supports all the metadata types in  `queryrouterd` and then some

1. Implement the strange UnionArray return type
2. Support filtering on sql_info name
3. Tests via JDBC
4. Tests via our own rust client (with filtering)

Follow on work is tracked in https://github.com/influxdata/influxdb_iox/issues/7253

### CODE REVIEW

File: flightsql/src/upstream/union_builder.rs [hunk 1]
```diff
@@ -0,0 +1,321 @@
+//! This is the same as https://github.com/apache/arrow-rs/blob/master/arrow-array/src/builder/union_builder.rs
```
Comment (alamb) [inline flightsql/src/upstream/union_builder.rs:1]: I don't think the UnionBuilder supports creating `Utf8` columns at the moment. I need to investigate some more

File: flightsql/src/sql_info/meta.rs [hunk 2]
```diff
@@ -0,0 +1,304 @@
+//! SQL metadata tables (originally from [queryrouterd])
+//!
+//! TODO: figure out how to generate these keywords automatically from DataFusion / sqlparser-rs
+//!
+//! [queryrouterd]: https://github.com/influxdata/idpe/blob/85aa7a52b40f173cc4d79ac02b3a4a13e82333c4/queryrouter/internal/server/flightsql_info.go#L4
+
+pub(crate) const SQL_INFO_SQL_KEYWORDS: &'static [&'static str] = &[
```
Comment (alamb) [inline flightsql/src/sql_info/meta.rs]: These lists are copied directly from `queryrouterd` which were good enough for Grafana

File: flightsql/src/sql_info/mod.rs [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+    }
+}
+
+#[allow(non_snake_case)]
+static INSTANCE: Lazy<SqlInfoList> = Lazy::new(|| {
+    // The following are not defined in the [`SqlInfo`], but are
+    // documented at
+    // https://arrow.apache.org/docs/format/FlightSql.html#protocol-buffer-definitions.
+
+    let SqlInfoFlightSqlServerSql = 4;
```
Comment (alamb) [inline flightsql/src/sql_info/mod.rs:104]: Here is the table that defines what metadata is reported

File: flightsql/src/sql_info/value.rs [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+/// Represents a dynamic value
+#[derive(Debug, Clone, PartialEq)]
+pub enum SqlInfoValue {
+    String(String),
+    Bool(bool),
+    BigInt(i64),
+    Bitmask(i32),
+    StringList(Vec<String>),
+    // TODO support more exotic metadata that requires the map of lists
+    //ListMap(BTreeMap<i32, Vec<i32>>),
```
Comment (alamb) [inline flightsql/src/sql_info/value.rs:22]: This is needed to support https://github.com/influxdata/influxdb_iox/issues/7253

File: flightsql/src/sql_info/value.rs [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+/// *              string_value: utf8,
+/// *              bool_value: bool,
+/// *              bigint_value: int64,
+/// *              int32_bitmask: int32,
+/// *              string_list: list<string_data: utf8>
+/// *              int32_to_int32_list_map: map<key: int32, value: list<$data$: int32>>
+/// * >
+///
+/// ```
+pub struct SqlInfoUnionBuilder {
```
Comment (alamb) [inline flightsql/src/sql_info/value.rs:93]: Here is a special version of `UnionBuilder` that I wrote after consultation with @tustvold

File: influxdb_iox/tests/jdbc_client/Main.java [hunk 6]
```diff
@@ -137,13 +137,157 @@ static void run_metadata(String url) throws SQLException {
...
+        //print_result_set(md.getColumns(null, null, null, null));
+
+        System.out.println("**************");
+        System.out.println("getFunctions:");
+        System.out.println("**************");
+        print_result_set(md.getFunctions(null, null, null));
+
+
+        // List from https://docs.oracle.com/javase/8/docs/api/java/sql/DatabaseMetaData.html
+        System.out.println("allProceduresAreCallable: " + md.allProceduresAreCallable());
```
Comment (alamb) [inline influxdb_iox/tests/jdbc_client/Main.java:154]: I wrote a python script to scraping the docs page and invoke all of these methods

File: influxdb_iox/tests/end_to_end_cases/flightsql.rs [hunk 7]
```diff
@@ -140,6 +140,80 @@ async fn flightsql_prepared_query() {
...
+                 {table_name},tag1=A,tag2=C val=43i 123457"
+            )),
+            Step::Custom(Box::new(move |state: &mut StepTestState| {
+                async move {
+                    let mut client = flightsql_client(state.cluster());
+
+                    // test with no filtering
+                    let batches = collect_stream(client.get_sql_info(vec![]).await.unwrap()).await;
+                    let total_rows: usize = batches.iter().map(|b| b.num_rows()).sum();
+                    assert_eq!(total_rows, 85);
```
Comment (appletreeisyellow) [inline influxdb_iox/tests/end_to_end_cases/flightsql.rs:167]: Does the number `85` come from [here](https://github.com/influxdata/influxdb_iox/blob/1cfaeee13f2331040637cf5aa90e3dd122e61323/flightsql/src/sql_info/mod.rs#L117-L237) where there are 85 `with_sql_info` values?

File: flightsql/src/sql_info/value.rs [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+/// Something that can be converted into u32 (the represenation of a
+/// [`SqlInfo`] name)
+pub trait SqlInfoName {
+    fn as_u32(&self) -> u32;
+}
+
+impl SqlInfoName for SqlInfo {
+    fn as_u32(&self) -> u32 {
+        // SqlInfos are u32 in the flight spec, but for some reason
+        // SqlInfo repr is an i32, so convert between them
```
[the remark this answers is not in the document]
Comment (alamb) [reply flightsql/src/sql_info/value.rs:65]: Thanks -- this really confused me at first too so I tried to leave a comment (largely for my future self) about what is going on.

File: flightsql/src/sql_info/value.rs [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+                self.int32_bitmask_values.append_value(*v);
+                (3, self.int32_bitmask_values.len())
+            }
+            SqlInfoValue::StringList(values) => {
+                // build list
+                for v in values {
+                    self.string_list_values.values().append_value(v);
+                }
+                // complete the list
+                self.string_list_values.append(true);
```
[the remark this answers is not in the document]
Comment (alamb) [reply flightsql/src/sql_info/value.rs:173]: What is going on is that this is creating a single  cell whose type is a list 

Tye `self.string_list_values.values().append_value()` is adding elements to each list and then `self.string_values.append(true)` is finishing up the list

So to create something like
```
string_list
---------
{one},
{},
{two, three}
```

The sequence of calls would be something like

```rust
// {one}
self.string_list_values.values().append_value("one");
self.string_list_values.append(true);
// {}
self.string_list_values.append(true);
// {two, three}
self.string_list_values.values().append_value("two");
self.string_list_values.values().append_value("three");
self.string_list_values.append(true);
```

I agree this is not clear -- and the docs for ListBuilder don't help much: https://docs.rs/arrow/35.0.0/arrow/array/struct.GenericListBuilder.html

File: influxdb_iox/tests/end_to_end_cases/flightsql.rs (hunk 7 above)
Comment (alamb) [reply influxdb_iox/tests/end_to_end_cases/flightsql.rs:167]: > Does the number 85 come from here where there are 85 with_sql_info values?

Yes, exactly. 

I will leave a comment to try and make that more explicit.

**Review by alamb — COMMENTED** [review]
Thank you for the review @appletreeisyellow

File: flightsql/src/sql_info/value.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (appletreeisyellow) [reply flightsql/src/sql_info/value.rs:173]: Thanks! The example makes sense to me 👍

[the remark this answers is not in the document]
Comment (alamb) [reply flightsql/src/sql_info/value.rs:173]: Upstream PR https://github.com/apache/arrow-rs/pull/3891

**Pull request merged**